Subscriptions: Change conditions for displaying modal

edanzer opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes #

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Adds extra checks and conditions for displaying (or not) the subscribe modal. Collectively these all avoid a situation where a user keeps re-seeing the subscribe modal even though they have subscribed or dismissed it.
* Do not show if user has already subscribed (`jetpack_blog_subscribe_` cookie is set).
* Do not show if `subscribe` query parameter is set (set when user first submits the subscribe form)
* If user dismisses the cookie, we add 1-day `jetpack-subscribe-modal-dismissed` cookie, and do not re-show the modal while that cookie exists.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1) Setup
   - Load this branch in your local Jetpack environment - you will need a jurassic tube connected site. 
   - Add the `add_filter( 'jetpack_subscriptions_modal_enabled', '__return_true', 20 );` filter. The feature is behind a feature flag, and this will enable it. You can add this to tools/docker/mu-plugins/0-sandbox.php, or temporarily in any file that loads on your instance. 
   - Go to Jetpack > Settings > Discussion > Newsletters and enable the "Enable subscriber modal". If you do not see that option enable "Let visitors subscribe to new posts and comments via email" and refresh the page.

2) Test dismissing the modal. 
   - Go to any single post on the front end, scroll, and confirm the modal shows. 
   - Dismiss the modal, refresh the page, and confirm the modal does not show again.
   - Open Dev Tools > Applications > Cookies, and confirm the `jetpack-subscribe-modal-dismissed` is set. 
   
3) Test the subscribe query parameter.
   - Delete the `jetpack-subscribe-modal-dismissed` cookie from the last test so the modal will show again. 
   - Add `?subscribe=anythinghere` to the url and refresh the page.
   - Confirm the modal never show.

4) Test subscribing to the blog.
   - Remove the subscribe query parameter from the last test.
   - Visit any single post, scroll down, add a comment, check the boxes to subscribe to the blog, and submit.
   - Once page reloads, confirm you see the `jetpack_blog_subscribe_...` cookie. 
   - If needed, delete the `jetpack-subscribe-modal-dismissed` cookie again. 
   - Refresh the page and confirm the modal never shows.

edanzer committed:
[not verified] Add conditions to hide subscribe modal

edanzer committed:
[not verified] changelog

edanzer commented:
I actually wonder if I should update to use this method in place of some of my current checks here: 
https://github.com/Automattic/jetpack/blob/5fdf5d7f2f534e1bc062eb5c37dc42c3f9b9d257/projects/plugins/jetpack/modules/subscriptions/views.php#L556

simison commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.js:
```diff
@@ -2,20 +2,22 @@ const { domReady } = wp;
 
+	const close = document.getElementsByClassName( 'jetpack-subscribe-modal__close' )[ 0 ];
+	const modalDismissedCookie = 'jetpack_subscribe_modal_dismissed';
+	const hasModalDismissedCookie = document.cookie.indexOf( modalDismissedCookie ) > -1;
```
I think `document.cookie` always exists, but I wanted to mention in case we need some protection, e.g. `document.cookie?.indexOf()`

edanzer replied:
I've added a check for this. For some reason, adding the optional chaining operator was giving an eslint error.

edanzer committed:
[not verified] Handle nonce verfication warning

edanzer committed:
Ensure document.cookie exists

edanzer committed:
Add one more check for subscribe status

edanzer commented:
Updated: I've added a check based on @simison's comment above, added an extra conditional check using the existing is_current_user_subscribed() method in views.php, and rebased.

lezama commented:
should we use this [cookie state class](https://github.com/Automattic/jetpack/blob/6040a756958a58ff13ca536810be259d0147dc54/projects/packages/status/src/class-cookiestate.php) somehow?

edanzer commented:
> should we use this [cookie state class](https://github.com/Automattic/jetpack/blob/6040a756958a58ff13ca536810be259d0147dc54/projects/packages/status/src/class-cookiestate.php) somehow?

Hey @lezama - Good question. Just spent some time looking at this. Interesting construct and seems like it could plausibly be used, but...
   - For setting cookies, we're only doing that once here, for a client-side-only 'dismissed' cookie, so the CookieState class is not as relevant for that. 
   - What could be interesting is if information about the user and user subscription status was stored using CookieState, you could key on that for checks and conditionals. But best I can tell, that information doesn't seem to be stored via CookieState. So we'd need to refactor other existing methods to store information that way first. If I'm understanding that right, it may not be worth adding that initially.

edanzer committed:
Add check for if user is member of site
